internal/proto/reader.go

```diff
@@ -92,240 +127,377 @@ func (r *Reader) readLine() ([]byte, error) {
+		var n int
+		n, err = replyLen(line)
 		if err != nil {
 			return nil, err
 		}
-		if m == nil {
-			err := fmt.Errorf("redis: got %.100q, but multi bulk parser is nil", line)
-			return nil, err
+		if a == nil {
+			return nil, fmt.Errorf("redis: got %.100q, but multi bulk parser is nil", line)
 		}
-		return m(r, n)
+		return a(r, line[0], int64(n))
 	}
 	return nil, fmt.Errorf("redis: can't parse %.100q", line)
 }
 
-func (r *Reader) ReadIntReply() (int64, error) {
-	line, err := r.ReadLine()
-	if err != nil {
-		return 0, err
-	}
-	switch line[0] {
-	case ErrorReply:
-		return 0, ParseErrorReply(line)
-	case IntReply:
-		return util.ParseInt(line[1:], 10, 64)
-	default:
-		return 0, fmt.Errorf("redis: can't parse int reply: %.100q", line)
+func (r *Reader) readFloat(line []byte) (float64, error) {
+	v := string(line[1:])
+	switch string(line[1:]) {
+	case "inf":
+		return uvinf, nil
+	case "-inf":
+		return uvneginf, nil
 	}
+	return strconv.ParseFloat(v, 64)
 }
 
-func (r *Reader) ReadString() (string, error) {
-	line, err := r.ReadLine()
-	if err != nil {
-		return "", err
-	}
-	switch line[0] {
-	case ErrorReply:
-		return "", ParseErrorReply(line)
-	case StringReply:
-		return r.readStringReply(line)
-	case StatusReply:
-		return string(line[1:]), nil
-	case IntReply:
-		return string(line[1:]), nil
-	default:
-		return "", fmt.Errorf("redis: can't parse reply=%.100q reading string", line)
+func (r *Reader) readBool(line []byte) (bool, error) {
+	switch string(line[1:]) {
+	case "t":
+		return true, nil
+	case "f":
+		return false, nil
 	}
+	return false, fmt.Errorf("redis: can't parse bool reply: %q", line)
 }
 
-func (r *Reader) readStringReply(line []byte) (string, error) {
-	if isNilReply(line) {
-		return "", Nil
+func (r *Reader) readBigInt(line []byte) (*big.Int, error) {
+	i := new(big.Int)
+	if i, ok := i.SetString(string(line[1:]), 10); ok {
+		return i, nil
 	}
+	return nil, fmt.Errorf("redis: can't parse bigInt reply: %q", line)
+}
 
-	replyLen, err := util.Atoi(line[1:])
+func (r *Reader) readStringReply(line []byte) (string, error) {
+	n, err := replyLen(line)
 	if err != nil {
 		return "", err
 	}
 
-	b := make([]byte, replyLen+2)
+	b := make([]byte, n+2)
 	_, err = io.ReadFull(r.rd, b)
 	if err != nil {
 		return "", err
 	}
 
-	return util.BytesToString(b[:replyLen]), nil
+	return util.BytesToString(b[:n]), nil
 }
 
-func (r *Reader) ReadArrayReply(m MultiBulkParse) (interface{}, error) {
-	line, err := r.ReadLine()
+func (r *Reader) readVerb(line []byte) (string, error) {
+	s, err := r.readStringReply(line)
+	if err != nil {
+		return "", err
+	}
+	if len(s) < 4 || s[3] != ':' {
+		return "", fmt.Errorf("redis: can't parse verbatim string reply: %q", line)
+	}
+	return s[4:], nil
+}
+
+// -------------------------------
+
+// Pathfinder find up possible errors in redis responses and discard attributes types,
+// `line` that returns the true reply.
+func (r *Reader) Pathfinder() ([]byte, error) {
```

Comment:
I am not sure I like the name `Pathfinder` - is this a common name that I don't recognize?

I think we should continue to call this `ReadLine` and introduce a new name later when we need to parse attributes.